crates/iroha_core/src/sumeragi/main_loop.rs

```diff
@@ -628,6 +628,16 @@ impl Sumeragi {
                 );
 
                 if let Ok(signatory_idx) = usize::try_from(signature.0) {
+                    if signatory_idx >= self.topology.as_ref().len() {
+                        error!(
+                            peer_id=%self.peer_id,
+                            role=%self.role(),
+                            ?signatory_idx,
+                            topology_size=%self.topology.as_ref().len(),
+                            "Signature index is out of range."
+                        );
+                        return;
+                    }
                     let signatory = &self.topology.as_ref()[signatory_idx];
```

Comment:
Out of bounds check is good, but it doesn't target the core issue - the data race in Sumeragi. The problem isn't really solved: imagine if the topology changed, but a certain index is still within the bounds - it wouldn't be an error, but Sumeragi will mess the peers between each other accidentally and can make incorrect decisions.

So, shouldn't there be a general solution for the data race condition instead?

Reply:
data races are inherent to distributed systems. It was incorrect to assume that received signatures are valid at this point. This can happen in regular sumeragi operation.